handlers/mist_triggers.go

```diff
@@ -0,0 +1,207 @@
+	StreamCache *StreamCache
+}
+
+// Trigger dispatches request to mapped method according to trigger name
+// Only single trigger callback is allowed on Mist.
+// All created streams and our handlers (segmenting, transcoding, et.) must share this endpoint.
+// If handler logic grows more complicated we may consider adding dispatch mechanism here.
+func (d *MistCallbackHandlersCollection) Trigger() httprouter.Handle {
+	return func(w http.ResponseWriter, req *http.Request, params httprouter.Params) {
+		t := &Trigger{w: w, req: req}
+		whichOne := req.Header.Get("X-Trigger")
+		switch whichOne {
+		case "PUSH_END":
+			d.TriggerPushEnd(t)
+		case "LIVE_TRACK_LIST":
+			d.TriggerLiveTrackList(t)
+		default:
+			errors.WriteHTTPBadRequest(w, "Unsupported X-Trigger", fmt.Errorf("unknown trigger '%s'", whichOne))
+			return
+		}
+	}
+}
+
+// TriggerLiveTrackList responds to LIVE_TRACK_LIST trigger.
+// It is stream-specific and must be blocking. The payload for this trigger is multiple lines,
+// each separated by a single newline character (without an ending newline), containing data:
+//   stream name
+//   push target URI
+// TriggerLiveTrackList is used only by transcoding.
+func (d *MistCallbackHandlersCollection) TriggerLiveTrackList(t *Trigger) {
+	payload, err := io.ReadAll(t.req.Body)
+	if err != nil {
+		t.Err("Cannot read payload", err)
+		return
+	}
+	lines := strings.Split(strings.TrimSuffix(string(payload), "\n"), "\n")
+	t.streamName = lines[0]
+	encodedTracks := lines[1]
+
+	yes, suffix := isTranscodeStream(t.streamName)
+	if !yes {
+		t.Err("unknown streamName format", nil)
+		return
+	}
+
+	if streamEnded := encodedTracks == "null"; streamEnded {
+		// SOURCE_PREFIX stream is no longer needed
+		inputStream := fmt.Sprintf("%s%s", SOURCE_PREFIX, suffix)
+		if err = d.MistClient.DeleteStream(inputStream); err != nil {
+			log.Printf("ERROR LIVE_TRACK_LIST DeleteStream(%s) %v", inputStream, err)
+		}
+		// Multiple pushes from RENDITION_PREFIX are in progress.
+		return
+	}
+	tracks := make(LiveTrackListTriggerJson)
+	if err = json.Unmarshal([]byte(encodedTracks), &tracks); err != nil {
+		t.Err("LiveTrackListTriggerJson json decode error", err)
+		return
+	}
+	// Start push per each video track
+	info, err := d.StreamCache.Transcoding.Get(t.streamName)
+	if err != nil {
+		t.Err("LIVE_TRACK_LIST unknown push source", err)
+		return
+	}
+	for i := range tracks { // i is generated name, not important, all info is contained in element
+		if tracks[i].Type != "video" {
+			// Only produce an rendition per each video track, selecting best audio track
+			continue
+		}
+		destination := fmt.Sprintf("%s/%s__%dx%d.ts?video=%d&audio=maxbps", info.UploadDir, t.streamName, tracks[i].Width, tracks[i].Height, tracks[i].Index) //.Id)
+		if err := d.MistClient.PushStart(t.streamName, destination); err != nil {
+			log.Printf("> ERROR push to %s %v", destination, err)
+		} else {
+			d.StreamCache.Transcoding.AddDestination(t.streamName, destination)
```

Comment:
Why if `PushStart()` returns an error, then we add a new destination? I think it's not clear 🤔

Reply:
Intent is opposite: no error (push started) then `AddDestination()`